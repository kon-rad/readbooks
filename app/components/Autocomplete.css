.suggestion {
    padding: 6px 8px;
    height: 24px;
    overflow: hidden;
    line-height: 16px;
}
.suggestion:hover {
    cursor: pointer;
    background-color: #b2cecf;
    color: #fffdef;
}

.selected {
    background-color: #b2cecf;
    color: #fffdef;
}

.autoCompList {
    position: absolute;
    background-color: #ffffff;
    margin: 0;
    padding: 0;
    list-style: none;
    right: 0;
    left: 0;
    overflow: hidden;
    height: auto;
    border: solid 1px grey;
    z-index: 99;
}