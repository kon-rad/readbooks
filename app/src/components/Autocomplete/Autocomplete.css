.suggestion {
    width: 278px;
    padding: 4px 8px;
    height: 24px;
    overflow: hidden;
    line-height: 16px;
}
.suggestion:hover {
    cursor: pointer;
    background-color: lightcyan;
}

.selected {
    background-color: lightcyan;
}

.autoCompList {
    position: absolute;
    background-color: #ffffff;
    margin: 0;
    padding: 0;
    list-style: none;
    right: 0;
    left: 0;
    overflow: hidden;
    height: auto;
    border: solid 1px grey;
}