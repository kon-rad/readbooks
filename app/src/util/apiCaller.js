
export default function bookSearch(query) {

  return fetch("https://www.googleapis.com/books/v1/volumes?q=" + encodeURIComponent(query))
    .then((res) => {
      return res.json();
    });
};
