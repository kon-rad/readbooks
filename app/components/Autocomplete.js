// @flow

import * as React from 'react';
import apiCaller from '../util/apiCaller';
import './Autocomplete.css';
import type { TitleResults } from '../types/BooksAPI';

type Props = {
  currentIndex: number,
  query: string,
  onClick: (hit: string) => void,
  onSuggestions: (hits: Array<TitleResults>) => void,
  open: boolean
};

type State = {
  hits: Array<TitleResults>
};

class Autocomplete extends React.Component<Props, State> {
  node = null;

  constructor(props: Props): void {
    super(props);
    this.state = {
      hits: []
    };
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClick, false);
  }

  componentDidUpdate(prevProps: Props): void {

    if (this.props.query !== prevProps.query) {
      if (this.props.query.trim() === '') {
        this.setState({ hits: [] });
        this.props.onSuggestions([]);
        return;
      }

      apiCaller(this.props.query).then(res => {
        this.setState({ hits: res.items || [] });
        this.props.onSuggestions(res.items);
      });
    }
  }

  handleClick = (e: MouseEvent): void => {
    if (this.node && e.target instanceof Node && this.node.contains(e.target)) {
      return;
    }

    this.setState({ hits: [] });
    this.props.onSuggestions([]);
  };

  renderHit = (hit: TitleResults): React.Element<'div'> => {
    let isSelected = false;
    if (this.props.currentIndex > -1)
      isSelected = this.state.hits[this.props.currentIndex].id === hit.id;
    const className = isSelected ? 'suggestion selected' : 'suggestion';
    const title = hit.volumeInfo.title;

    return (
      <div
        key={hit.id}
        role="option"
        onClick={() => this.props.onClick(hit.volumeInfo.title)}
        aria-selected={isSelected}
        className={className}
      >
        {title}
      </div>
    );
  };

  render() {
    const style = {};
    if (
      !this.props.open ||
      !this.state.hits ||
      (this.state.hits && this.state.hits.length === 0)
    ) {
      style.display = 'none';
    }
    return (
      <div
        ref={node => (this.node = node)}
        role="list"
        className="autoCompList"
        style={style}
      >
        {this.state.hits.map(this.renderHit)}
      </div>
    );
  }
}

export default Autocomplete;
