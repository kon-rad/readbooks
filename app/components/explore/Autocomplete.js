// @flow
import * as React from 'react';
import apiCaller from 'util/apiCaller';
import 'components/explore/Autocomplete.css';
import type { SearchResults } from 'types/Posts';

type Props = {
  currentIndex: number,
  query: string,
  onClick: (hit: string) => void,
  onSuggestions: (hits: Array<SearchResults>) => void,
  open: boolean
};

type State = {
  hits: Array<SearchResults>
};

class Autocomplete extends React.Component<Props, State> {
  node = null;

  constructor(props: Props): void {
    super(props);
    this.state = {
      hits: []
    };
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClick);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClick);
  }

  componentDidUpdate(prevProps: Props): void {
    if (this.props.query !== prevProps.query) {
      if (this.props.query.trim() === '') {
        this.setState({ hits: [] });
        this.props.onSuggestions([]);
        return;
      }

      apiCaller(this.props.query).then(res => {
        this.setState({ hits: res.items || [] });
        this.props.onSuggestions(res.items);
      });
    }
  }

  handleClick = (e: MouseEvent): void => {
    if (this.node && e.target instanceof Node && this.node.contains(e.target)) {
      return;
    }

    this.setState({ hits: [] });
    this.props.onSuggestions([]);
  };

  renderHit = (hit: SearchResults): React.Element<'div'> => {
    const isSelected = (this.props.currentIndex > -1) 
      ? this.state.hits[this.props.currentIndex].id === hit.id 
      : false;

    const className = isSelected ? 'suggestion selected' : 'suggestion';
    const title = hit.volumeInfo.title;

    return (
      <div
        key={hit.id}
        role="option"
        onClick={() => this.props.onClick(hit.volumeInfo.title)}
        aria-selected={isSelected}
        className={className}
      >
        {title}
      </div>
    );
  };

  render() {
    const autocompleteStyle = {
      display: (!this.props.open || this.state.hits.length === 0) ? 'none' : 'block'
    };

    return (
      <div
        id="autoCompList"
        ref={node => (this.node = node)}
        role="list"
        className="autoCompList"
        style={autocompleteStyle}
      >
        {this.state.hits.map(this.renderHit)}
      </div>
    );
  }
}

export default Autocomplete;
