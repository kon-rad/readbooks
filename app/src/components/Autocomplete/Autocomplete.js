

import React from 'react';
import apiCaller from '../../util/apiCaller';
import styles from './Autocomplete.css';
import classNames from 'classnames/bind';

class Autocomplete extends React.Component {

  cx = classNames.bind(styles);

  constructor(props) {
    super(props);
    this.state = {
      hits: [],
    };
    console.log(styles);
  }

  componentDidUpdate(prevProps) {
    if (this.props.query !== prevProps.query) {

      if (this.props.query.trim() === '') {
        this.setState({ hits: [] });
        this.props.onSuggestions([]);
        return;
      }

      apiCaller(this.props.query)
        .then((res) => {
          this.setState({ hits: res.items || [] });
          this.props.onSuggestions(res.items);
        })
    }
  }

  renderHit = hit => {
    let isSelected = false;
    if (this.props.currentIndex > -1) isSelected = this.state.hits[this.props.currentIndex].id === hit.id;
    let className = this.cx({
      suggestion: true,
      selected: isSelected
    });
    let title = hit.volumeInfo.title;
    if (title.length > 34) {
      title = title.slice(0,34);
      title = title.split(' ').slice(0, -1).join(' ') + ' ...';
    }

    return (
      <div key={hit.id}
           role="option"
           onClick={() => this.props.onClick(hit.volumeInfo.title)}
           aria-selected={isSelected}
           className={className}
      >
        {title}
      </div>
    )
  };

  render() {
    const style = {};
    if (!this.props.open || this.state.hits.length === 0) {
      style.display = 'none';
    }
    return (
      <div role="list" className={styles.autoCompList} style={style}>
        {this.state.hits.map(this.renderHit)}
      </div>
    )
  }
}

export default Autocomplete;