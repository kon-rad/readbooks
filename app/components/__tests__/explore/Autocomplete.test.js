import React from 'react';
import Autocomplete from 'components/explore/Autocomplete';
import { shallow, mount } from 'enzyme';
import apiCaller from 'util/apiCaller';

jest.mock('util/apiCaller', () => jest.fn());

const mountComponent = (currentIndex, query, onClick, onSuggestions, open) => {
  return mount(
    <Autocomplete
      currentIndex={currentIndex}
      query={query}
      onClick={onClick}
      onSuggestions={onSuggestions}
      open={open}
    />
  );
};

describe('Autocomplete', () => {
  it('renders matching snapshot', () => {
    const wrapper = shallow(<Autocomplete />);
    expect(wrapper).toMatchSnapshot();
  });

  let hits;
  let mockAPI;
  beforeEach(() => {
    hits = [
      {
        id: 'resultID1',
        volumeInfo: {
          title: 'Title 1'
        }
      },
      {
        id: 'resultID2',
        volumeInfo: {
          title: 'Title 2'
        }
      }
    ];

    mockAPI = apiCaller.mockImplementation(() =>
      Promise.resolve({ items: hits })
    );
  });
  afterEach(() => {
    jest.resetAllMocks()
  });

  it('should perform search and display a list when a new query is entered', async () => {
    const onSuggestions = jest.fn();
    const wrapper = mountComponent(
      -1,
      '',
      jest.fn(),
      onSuggestions,
      false
    );
    wrapper.setProps({ query: 'new query' });
    expect(apiCaller).toHaveBeenCalledWith('new query');
    await mockAPI();
    expect(wrapper.state('hits')).toEqual(hits);
    expect(onSuggestions).toHaveBeenCalledWith(hits);
    wrapper.unmount();
  });

  it('should not search when the query trims to empty string', async () => {

    const onSuggestions = jest.fn();
    const wrapper = mountComponent(
      -1,
      '',
      jest.fn(),
      onSuggestions,
      false
    );

    wrapper.setProps({ query: '    ' });
    await mockAPI();
    expect(wrapper.state('hits')).toEqual([]);
    expect(onSuggestions).toHaveBeenCalledWith([]);
    expect(mockAPI.mock.calls).toEqual([[]]);
    wrapper.unmount();
  });

  it('should not search when new query is empty after previous search', async () => {

    const onSuggestions = jest.fn();
    const wrapper = mountComponent(
      -1,
      'existing query',
      jest.fn(),
      onSuggestions,
      false
    );

    wrapper.setProps({ query: '' });
    await mockAPI();
    expect(wrapper.state('hits')).toEqual([]);
    expect(onSuggestions).toHaveBeenCalledWith([]);
    expect(mockAPI.mock.calls).toEqual([[]]);
    wrapper.unmount();
  });

  it('should call renderHit for each hit when query is entered', async () => {
    const onSuggestions = jest.fn();
    const mockRenderHit = jest.fn();
    const wrapper = mountComponent(
      -1,
      '',
      jest.fn(),
      onSuggestions,
      false
    );
    wrapper.instance().renderHit = mockRenderHit;
    wrapper.setProps({ query: 'new query' });
    await mockAPI();

    expect(mockRenderHit).toHaveBeenCalledTimes(2);
    wrapper.unmount();
  });

  it('should render an item for each hit when query is entered', async () => {
    const onSuggestions = jest.fn();
    const wrapper = mountComponent(
      -1,
      '',
      jest.fn(),
      onSuggestions,
      false
    );
    wrapper.setProps({ query: 'new query' });
    await mockAPI();

    expect(wrapper.find('div')).toHaveLength(1);
    expect(wrapper.state('hits')).toEqual(hits);
    wrapper.unmount();
  });
});
