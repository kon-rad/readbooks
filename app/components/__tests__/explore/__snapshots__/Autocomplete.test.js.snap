// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete renders matching snapshot 1`] = `
<div
  className="autoCompList"
  role="list"
  style={
    Object {
      "display": "none",
    }
  }
/>
`;
